File: prefPane/update.m
/*
 * id-updater
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#import "update.h"

#import <CryptoTokenKit/CryptoTokenKit.h>
#import <Security/Security.h>

#include <sys/utsname.h>

#include "config.h"

#define UPDATER_ID @"ee.ria.id-updater"

@implementation Update {
    NSMutableURLRequest *request;
    NSString *signature;
}

- (id)initWithDelegate:(id <UpdateDelegate>)delegate {
    if (self = [super init]) {
        self.delegate = delegate;
        self.updaterversion = [self versionInfo:@"ee.ria.ID-updater"];
        self.baseversion = [self versionInfo:@"ee.ria.open-eid"];
        self.clientversion = [self versionInfo:@"ee.ria.qdigidocclient"];
        self.digidoc4 = [self versionInfo:@"ee.ria.qdigidoc4"];
        self.utilityversion = [self versionInfo:@"ee.ria.qesteidutil"];
    }
    return self;
}

- (BOOL)checkCertificatePinning:(NSURLAuthenticationChallenge *)challenge {
    SecTrustRef serverTrust = challenge.protectionSpace.serverTrust;
    SecTrustResultType trustResult;
    SecTrustEvaluate(serverTrust, &trustResult);
    if ((trustResult == kSecTrustResultUnspecified ||
         trustResult == kSecTrustResultProceed) &&
        SecTrustGetCertificateCount(serverTrust) > 0) {
        SecCertificateRef certificate = SecTrustGetCertificateAtIndex(serverTrust, 0);
        NSData *der = CFBridgingRelease(SecCertificateCopyData(certificate));
        return [self.cert_bundle containsObject:der];
    }
    return NO;
}

- (void)request {
    NSURL *url = [NSURL URLWithString:@CONFIG_URL];
    url = [url.URLByDeletingLastPathComponent URLByAppendingPathComponent:@"config.rsa"];
    request = [NSMutableURLRequest requestWithURL:url
        cachePolicy:NSURLRequestReloadIgnoringLocalCacheData timeoutInterval:10];
    [request addValue:[self userAgent:YES] forHTTPHeaderField:@"User-Agent"];
    [[NSURLSession.sharedSession dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        [self receivedData:data withResponse:response];
    }] resume];
}

- (NSString*)userAgent:(BOOL)diangostics {
    NSDictionary *os = [NSDictionary dictionaryWithContentsOfFile:@"/System/Library/CoreServices/SystemVersion.plist"];
    struct utsname unameData;
    uname(&unameData);

    NSString *devices = [TKSmartCardSlotManager.defaultManager.slotNames componentsJoinedByString:@"/"];
    NSMutableArray *agent = [NSMutableArray arrayWithObject:[NSString stringWithFormat:@"id-updater/%@", self.updaterversion]];
    if (diangostics && self.clientversion.length) {
        [agent addObject:[NSString stringWithFormat:@"qdigidocclient/%@", self.clientversion]];
    }
    if (diangostics && self.utilityversion.length) {
        [agent addObject:[NSString stringWithFormat:@"qesteidutility/%@", self.utilityversion]];
    }
    if (diangostics && self.digidoc4.length) {
        [agent addObject:[NSString stringWithFormat:@"qdigidoc4/%@", self.digidoc4]];
    }
    NSString *locale = NSBundle.mainBundle.preferredLocalizations[0];
    [agent addObject:[NSString stringWithFormat:@"(Mac OS %@ (%lu/%s)) Locale: %@ / %@ Devices: %@",
        os[@"ProductVersion"], sizeof(void *)<<3, unameData.machine, locale, @"UTF-8", devices]];
    return [agent componentsJoinedByString:@" "];
}

- (BOOL)verify:(NSData *)data error:(NSError **)error {
    NSString *pem = @((char*)config_pub);
    pem = [pem stringByReplacingOccurrencesOfString:@"-----BEGIN RSA PUBLIC KEY-----" withString:@""];
    pem = [pem stringByReplacingOccurrencesOfString:@"-----END RSA PUBLIC KEY-----" withString:@""];
    pem = [NSString stringWithFormat:@"%@%@", @"MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8A", pem];
    NSData *keyData = [[NSData alloc] initWithBase64EncodedString:pem options:NSDataBase64DecodingIgnoreUnknownCharacters];
    NSDictionary *parameters = @{
        (__bridge id)kSecAttrKeyType: (__bridge id)kSecAttrKeyTypeRSA,
        (__bridge id)kSecAttrKeyClass: (__bridge id)kSecAttrKeyClassPublic
    };
    CFErrorRef err = nil;
    id key = CFBridgingRelease(SecKeyCreateFromData((__bridge CFDictionaryRef)parameters, (__bridge CFDataRef)keyData, &err));
    if (err) { if(error) *error = CFBridgingRelease(err); return false; }

    NSData *signatureData = [[NSData alloc] initWithBase64EncodedString:signature options:NSDataBase64DecodingIgnoreUnknownCharacters];
    BOOL isValid = SecKeyVerifySignature((__bridge SecKeyRef)key, kSecKeyAlgorithmRSASignatureMessagePKCS1v15SHA512,
                                         (__bridge CFDataRef)data, (__bridge CFDataRef)signatureData, &err);
    if (err) { if(error) *error = CFBridgingRelease(err); return false; }
    return isValid;
}

- (BOOL)verifyCMSSignature:(NSData *)signatureData data:(NSData *)data cert:(NSData *)cert {
    #define RETURN_IF_OERROR(MSG) if (oserr) { NSLog(MSG); return false; }
    CMSDecoderRef decoderRef;
    OSStatus oserr = CMSDecoderCreate(&decoderRef);
    RETURN_IF_OERROR(@"CMSDecoderCreate")
    id decoder = CFBridgingRelease(decoderRef);

    oserr = CMSDecoderUpdateMessage((__bridge CMSDecoderRef)decoder, signatureData.bytes, signatureData.length);
    RETURN_IF_OERROR(@"CMSDecoderUpdateMessage")
    oserr = CMSDecoderFinalizeMessage((__bridge CMSDecoderRef)decoder);
    RETURN_IF_OERROR(@"CMSDecoderFinalizeMessage")
    oserr = CMSDecoderSetDetachedContent((__bridge CMSDecoderRef)decoder, (__bridge CFDataRef)data);
    RETURN_IF_OERROR(@"CMSDecoderSetDetachedContent")

    size_t numSignersOut = 0;
    oserr = CMSDecoderGetNumSigners((__bridge CMSDecoderRef)decoder, &numSignersOut);
    RETURN_IF_OERROR(@"CMSDecoderGetNumSigners")
    if (numSignersOut != 1) {
        NSLog(@"Invalid number of signers: %lu", numSignersOut);
        return false;
    }

    SecPolicyRef policy = SecPolicyCreateBasicX509();
    CMSSignerStatus status;
    oserr = CMSDecoderCopySignerStatus((__bridge CMSDecoderRef)decoder, 0, policy, TRUE, &status, nil, nil);
    CFRelease(policy);
    RETURN_IF_OERROR(@"CMSDecoderCopySignerStatus")
    bool isValid = status == kCMSSignerValid;

    SecCertificateRef signerCert;
    oserr = CMSDecoderCopySignerCert((__bridge CMSDecoderRef)decoder, 0, &signerCert);
    RETURN_IF_OERROR(@"CMSDecoderCopySignerCert")
    bool isSameCert = [cert isEqualToData:CFBridgingRelease(SecCertificateCopyData(signerCert))];
    CFRelease(signerCert);

    NSLog(@"Signature is (%d) and cert is equal(%d)", isValid, isSameCert);
    return isValid && isSameCert;
}

- (NSString*)versionInfo:(NSString *)pkg {
    NSDictionary *list = [NSDictionary dictionaryWithContentsOfFile:[NSString stringWithFormat:@"/var/db/receipts/%@.plist", pkg]];
    return list ? list[@"PackageVersion"] : [NSString string];
}

- (void)URLSession:(NSURLSession *)session didReceiveChallenge:(NSURLAuthenticationChallenge *)challenge
        completionHandler:(void (^)(NSURLSessionAuthChallengeDisposition, NSURLCredential *))completionHandler {
    if([challenge.protectionSpace.authenticationMethod isEqualToString:NSURLAuthenticationMethodServerTrust]) {
        if ([self checkCertificatePinning:challenge]) {
            completionHandler(NSURLSessionAuthChallengeUseCredential,
                              [NSURLCredential credentialForTrust:challenge.protectionSpace.serverTrust]);
        } else {
            completionHandler(NSURLSessionAuthChallengeCancelAuthenticationChallenge, nil);
        }
    }
}

- (void)receivedData:(NSData *)data withResponse:(NSURLResponse *)response {
    NSHTTPURLResponse *http = (NSHTTPURLResponse*)response;
    if (http.statusCode != 200) {
        [self.delegate didFinish:[NSError errorWithDomain:@"ee.ria.ID-updater" code:FileNotFound userInfo:nil]];
        return;
    }

    NSString *file = request.URL.absoluteString.lastPathComponent;
    NSError *error;
    if ([file isEqualToString:@"config.json"]) {
        if (![self verify:data error:&error]) {
            NSLog(@"Verify error: %@", error);
            [self.delegate didFinish:[NSError errorWithDomain:@"ee.ria.ID-updater" code:InvalidSignature userInfo:nil]];
            return;
        }
        NSError *error = nil;
        NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
        if (!json) {
            [self.delegate didFinish:error];
            return;
        }

        NSDateFormatter *df = [[NSDateFormatter alloc] init];
        df.dateFormat = @"yyyyMMddHHmmss'Z'";
        df.timeZone = [NSTimeZone timeZoneForSecondsFromGMT:0];
        if ([NSDate.date compare:[df dateFromString:json[@"META-INF"][@"DATE"]]] == NSOrderedAscending) {
            [self.delegate didFinish:[NSError errorWithDomain:@"ee.ria.ID-updater" code:DateLaterThanCurrent userInfo:nil]];
            return;
        }

        self.centralConfig = json;
        NSMutableArray *certs = [NSMutableArray arrayWithCapacity:[self.centralConfig[@"CERT-BUNDLE"] count]];
        for (NSString *b64 in self.centralConfig[@"CERT-BUNDLE"]) {
            [certs addObject:[[NSData alloc] initWithBase64EncodedString:b64 options:NSDataBase64DecodingIgnoreUnknownCharacters]];
        }
        self.cert_bundle = certs;
        NSString *version = json[@"OSX-LATEST"];
        if (version) {
            NSLog(@"Remote version: %@", version);
            if ([version compare:self.baseversion options:NSNumericSearch] > 0) {
                [self.delegate updateAvailable:version filename:json[@"OSX-DOWNLOAD"]];
            }
        }
        NSString *message_url = json[@"UPDATER-MESSAGE-URL"];
        if (message_url) {
            NSURLSession *session = [NSURLSession sessionWithConfiguration:NSURLSessionConfiguration.defaultSessionConfiguration delegate:self delegateQueue:Nil];
            NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:message_url]
                cachePolicy:NSURLRequestReloadIgnoringLocalCacheData timeoutInterval:10];
            [request addValue:[self userAgent:NO] forHTTPHeaderField:@"User-Agent"];
            [[session dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                NSHTTPURLResponse *http = (NSHTTPURLResponse*)response;
                if (http.statusCode != 200) {
                    [self.delegate didFinish:nil];
                    return;
                }
                [self.delegate message:[[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]];
                [self.delegate didFinish:nil];
            }] resume];
            return;
        }
        NSString *message = json[@"OSX-MESSAGE"];
        if (message) {
            NSLog(@"Message: %@", message);
            [self.delegate message:message];
        }
        [self.delegate didFinish:error];
    } else if ([file isEqualToString:@"config.rsa"]) {
        signature = [[NSString alloc] initWithData:data encoding:NSASCIIStringEncoding];
        request.URL = [NSURL URLWithString:@CONFIG_URL];
        [[NSURLSession.sharedSession dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
            [self receivedData:data withResponse:response];
        }] resume];
    }
}

@end

File: prefPane/update.swift
// SPDX-FileCopyrightText: Estonian Information System Authority
// SPDX-License-Identifier: LGPL-2.1-or-later

import CryptoTokenKit

#if !NO_CONFIG_MODULE
import config
#endif

@objc public enum UpdateError: Int, Error {
    case invalidSignature = 1000
    case dateLaterThanCurrent = 1001
    case fileNotFound = 1002
    case jsonError = 1004
}

@objc public protocol UpdateDelegate: AnyObject {
    func didFinish(_ error: Error?)
    func message(_ message: String)
    func updateAvailable(_ available: String, filename: URL)
}

open class Update: NSObject, URLSessionDelegate {
    public var delegate: UpdateDelegate?
    @objc public var baseVersion: String? { Update.versionInfo("ee.ria.open-eid") }
    @objc(cert_bundle) public var certBundle: [Data] = []

    private let url: URL
    private let key: SecKey
    private let algorithm: SecKeyAlgorithm
    private let decoder = {
        let df = DateFormatter()
        df.dateFormat = "yyyyMMddHHmmss'Z'"
        df.timeZone = TimeZone(secondsFromGMT: 0)
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .formatted(df)
        return decoder
    }()

    @objc
    public init?(delegate: UpdateDelegate?) {
        guard let url = URL(string: CONFIG_URL) else {
            return nil
        }
        let pem = withUnsafeBytes(of: config_ecpub) {
            String(decoding: $0.bindMemory(to: UInt8.self), as: UTF8.self)
        }
        let keyData = Data(base64Encoded: pem
            .replacingOccurrences(of: "-----BEGIN PUBLIC KEY-----", with: "")
            .replacingOccurrences(of: "-----END PUBLIC KEY-----", with: ""), options: .ignoreUnknownCharacters)!
        guard let record = TKBERTLVRecord(from: keyData),
              let eckey = TKBERTLVRecord
            .sequenceOfRecords(from: record.value)?
            .first(where: { $0.tag == 0x03 })?
            .value.dropFirst() else {
            NSLog("Failed to parse key")
            return nil
        }
        let parameters = [
            kSecAttrKeyType: kSecAttrKeyTypeEC,
            kSecAttrKeyClass: kSecAttrKeyClassPublic,
        ]
        var err: Unmanaged<CFError>?
        guard let key = SecKeyCreateWithData(eckey as CFData, parameters as CFDictionary, &err) else {
            NSLog("Failed to create key: \(String(describing: err))")
            return nil
        }
        guard let attributes = SecKeyCopyAttributes(key) as? [CFString: Any],
            let keySize = attributes[kSecAttrKeySizeInBits] as? Int else {
            NSLog("Failed to get key size")
            return nil
        }
        self.delegate = delegate
        self.key = key
        self.url = url.deletingLastPathComponent()
        switch keySize {
        case 256: algorithm = .ecdsaSignatureMessageX962SHA256
        case 384: algorithm = .ecdsaSignatureMessageX962SHA384
        default:  algorithm = .ecdsaSignatureMessageX962SHA512
        }
    }

    func checkCertificatePinning(_ challenge: URLAuthenticationChallenge) -> Bool {
        guard let serverTrust = challenge.protectionSpace.serverTrust else { return false }
        var error: CFError?
        if !SecTrustEvaluateWithError(serverTrust, &error) {
            NSLog("SecTrustEvaluateWithError \(String(describing: error))")
            //return false
        }

        var trustResult = SecTrustResultType.invalid
        SecTrustGetTrustResult(serverTrust, &trustResult)
        if (trustResult == .unspecified ||
           trustResult == .proceed ||
           trustResult == .recoverableTrustFailure),
           let chain = SecTrustCopyCertificateChain(serverTrust) as? [SecCertificate],
           let firstCert = chain.first {
            return certBundle.contains(SecCertificateCopyData(firstCert) as Data)
        }
        return false
    }

    @objc(request) public func makeRequest() {
        Task {
            do {
                var request = URLRequest(url: url.appendingPathComponent("config.ecc"), cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 10)
                request.addValue(userAgent(diagnostics: true), forHTTPHeaderField: "User-Agent")
                var (data, response) = try await URLSession.shared.data(for: request)
                guard let httpResponse = response as? HTTPURLResponse,
                      httpResponse.statusCode == 200,
                      !data.isEmpty else {
                    self.delegate?.didFinish(UpdateError.fileNotFound)
                    return
                }
                guard let signature = Data(base64Encoded: data, options: .ignoreUnknownCharacters) else {
                    NSLog("Invalid signature encoding")
                    self.delegate?.didFinish(UpdateError.invalidSignature)
                    return
                }
                request.url = url.appendingPathComponent("config.json")
                (data, response) = try await URLSession.shared.data(for: request)
                guard let httpResponse = response as? HTTPURLResponse,
                      httpResponse.statusCode == 200,
                      !data.isEmpty else {
                    self.delegate?.didFinish(UpdateError.fileNotFound)
                    return
                }
                await self.receivedData(data: data, signature: signature)
            } catch {
                NSLog("Failed to fetch file \(error)")
                self.delegate?.didFinish(UpdateError.fileNotFound)
            }
        }
    }

    private func receivedData(data: Data, signature: Data) async {
        var error: Unmanaged<CFError>?
        guard SecKeyVerifySignature(key, algorithm, data as CFData, signature as CFData, &error) else {
            NSLog("Verify error: \(error?.takeRetainedValue() as Error?)")
            delegate?.didFinish(UpdateError.invalidSignature)
            return
        }
        let config: Config
        do {
            config = try decoder.decode(Config.self, from: data)
        } catch {
            NSLog("Failed to parse json: \(error)")
            delegate?.didFinish(UpdateError.jsonError)
            return
        }
        guard config.metaInfo.DATE < Date() else {
            NSLog("Failed to parse json: \(config.metaInfo.DATE) < \(Date())")
            delegate?.didFinish(UpdateError.dateLaterThanCurrent)
            return
        }

        NSLog("Config: \(config.metaInfo.SERIAL) \(config.metaInfo.URL) \(config.metaInfo.DATE)")
        certBundle = config.certBundle

        NSLog("Remote version: \(config.osxLatest) base version: \(baseVersion ?? "nil")")
        if config.osxLatest.compare(baseVersion ?? "", options: .numeric) == .orderedDescending {
            delegate?.updateAvailable(config.osxLatest, filename: config.osxDownload)
        }

        if let messageURL = config.updaterMessageURL {
            do {
                var request = URLRequest(url: messageURL, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 10)
                request.addValue(userAgent(diagnostics: false), forHTTPHeaderField: "User-Agent")
                let session = URLSession(configuration: .default, delegate: self, delegateQueue: nil)
                let (data, response) = try await session.data(for: request)
                if let httpResponse = response as? HTTPURLResponse,
                   httpResponse.statusCode == 200,
                   let message = String(data: data, encoding: .utf8) {
                    self.delegate?.message(message)
                }
            } catch {
                NSLog("Failed to fetch message \(error)")
            }
        }

        if let message = config.osxMessage {
            NSLog("Message: \(message)", message)
            delegate?.message(message)
        }
        delegate?.didFinish(nil)
    }

    public func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        if challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
            checkCertificatePinning(challenge) {
            completionHandler(.useCredential, URLCredential(trust: challenge.protectionSpace.serverTrust!))
        } else {
            completionHandler(.cancelAuthenticationChallenge, nil)
        }
    }

    @objc(userAgent:) public func userAgent(diagnostics: Bool) -> String {
        let os = NSDictionary(contentsOfFile: "/System/Library/CoreServices/SystemVersion.plist") as? [String: Any]
        var systemInfo = utsname()
        uname(&systemInfo)
        let machine = withUnsafeBytes(of: systemInfo.machine) {
            String(cString: $0.bindMemory(to: CChar.self).baseAddress!)
        }
        let devices = TKSmartCardSlotManager.default?.slotNames.joined(separator: "/") ?? ""
        var agent = ["id-updater/\(Update.versionInfo("ee.ria.ID-updater") ?? "0.0.0.0")"]

        if diagnostics, let digidoc4 = Update.versionInfo("ee.ria.qdigidoc4"), !digidoc4.isEmpty {
            agent.append("qdigidoc4/\(digidoc4)")
        }

        let locale = Locale.current.identifier
        agent.append("(Mac OS \(os?["ProductVersion"] as? String ?? "0.0.0") (\(MemoryLayout<UInt>.size * 8)/\(machine)) Locale: \(locale) / UTF-8 Devices: \(devices)")

        return agent.joined(separator: " ")
    }

    static func versionInfo(_ pkg: String) -> String? {
        let list = NSDictionary(contentsOfFile: "/var/db/receipts/\(pkg).plist")
        return list?["PackageVersion"] as? String
    }
}

struct Config: Decodable {
    struct MetaInfo: Decodable {
        let VER: Int
        let SERIAL: Int
        let URL: URL
        let DATE: Date
    }

    let metaInfo: MetaInfo
    let osxDownload: URL
    let osxLatest: String
    let osxMessage: String?
    let updaterMessageURL: URL?
    let certBundle: [Data]

    private enum CodingKeys: String, CodingKey {
        case metaInfo = "META-INF"
        case osxDownload = "OSX-DOWNLOAD"
        case osxLatest = "OSX-LATEST"
        case osxMessage = "OSX-MESSAGE"
        case updaterMessageURL = "UPDATER-MESSAGE-URL"
        case certBundle = "CERT-BUNDLE"
    }
}
